api_pagopa.yaml

```diff
@@ -0,0 +1,289 @@
+swagger: "2.0"
+info:
+  version: 1.0.0
+  title: Proxy PagoPA
+  description: Proxy for PagoPA payment services
+basePath: /api/v1
+schemes:
+  - https
+paths:
+  "/payment/check":
+    x-swagger-router-controller: PaymentController
+    get:
```

Comment:
it looks like it's a POST (has body payload) BUT as it's a read only op, we better use something like `GET payment/{CF}/{CodiceAvviso}` IMHO, cc @cloudify

Reply:
The payload is:
CodiceIdRPT: 
 CF: FiscalCode,
  AuxDigit:  0 | 1 | 2 | 3
  CodIUV: PatternString("[0-9]{15}|[0-9]{17}")
  CodStazPA?: PatternString("[0-9]{2}") // optional
});

If you prefer, we can move it on the URL but I think it's more readable and clear the current solution. Let's see Federico opinion :)

Reply:
that's right but CodiceAvviso= AuxDigit+CodIUV+CodStazPA (= payment id for this fiscal code)

Reply:
I think you both have a point:

1. in REST semantics, a GET operation will synchronously retrieve a certain resource and the resource ID should be in the path (e.g. `GET /users/gunzip`)
1. in our case the resource ID is a complex object to in some sense it would make sense to serialize it as json

Now, let me show you a different perspective:

* before calling `checkPayment`, the app needs to call `activatePayment`
* by calling `activatePayment`, a kind of "transaction" gets created (if I recall correctly here the proxy sends the request to pagopa along with a unique `operationId` for correlating the later async response from pagopa) - so, in effect, `operationId` is the identifier of this transaction
* afterwards, when the app calls `checkPayment`, it's in fact checking the status of this transaction (is the transaction completed?)

So my suggestion is that:

* `activatePayment` to be a PUT operation, with the payment request ID in the body
* `activatePayment` to return the `operationId`
* `checkPayment` to be a GET operation with the `operationId` in the path and no body

Pros: reflects the semantic of the async logic behind the REST API
Cons: the client needs to keep track of the `operationId` (this is less convenient for the client but I think it makes sense to link the check to the actual activate operation, making the whole flow more correct)

Reply:
Another option (aligned to what @gunzip suggests) would be to define a canonical representation of the payment request ID (*) that we can use as resource identifiers.

(*) that would basically be similar to the string contained in the QR code

Reply:
that's right but we do have a transaction id (codiceContestoPagamento) so why not use that one (see below, GET "/payment/activation/{codiceContestoPagamento}")

Reply:
Mmh wait, I see now that we have `checkPaymentActivationStatusFromDB` below that is already what I meant for `checkPayment` - I see now, sorry I got confused - is `checkPayment` the _verifica_ operation? perhaps better call it `verifyPaymentRequest` or something like that, because we're dealing with a payment request (payment has not started yet).

In this case I think the more sensible approach is to rely on a canonical representation of the payment request ID as suggested by @gunzip

Reply:
I'm more inclined towards a POST (than a PUT) as usually when you do PUT /something/{id}
you expect to get it with GET /something/{id}but that won't be our case (as the first id is anotehr thing from the transaction id returned)

Reply:
Ok, definitely the flow is:
GET /payment/{CF}/{CodiceAvviso}      // Verifica RPT
POST /payment/activation      // Attiva RPT
GET /payment/activation/:codiceContestoPagamento     // Check Attiva RPT async result

codiceContestPagamento is exactly what you called operationId and it's generated by proxyPagoPa on the POST request

CodiceAvviso will be the concatenation of AuxDigit, CodIUV and CodStazPA and it will be parsed and splitted by pagoPaProxy (pagoPa SOAP requires separated values)

Reply:
@gunzip you are right:`POST` creates a new resource and you get an ID back (e.g. `POST /users`) while `PUT` upserts a resource by ID (e.g. `PUT /users/gunzip`). As we are creating an activation request, receiving the `codiceContestoPagamento` back, this is a `POST`.

Reply:
@lskoder yes correct, a couple of additional points:

* `CodiceAvviso` is already unique (contains the fiscal code of the administration) so I would avoid having also the CF in the path - also since the `CodiceAvviso` identifies a payment request (not an actual payment transaction) I would call the resource `payment-requests` (resources should be in plural form)
* in the "attiva" flow, we're dealing with _payment activation_ resources/transactions so I suggest to name the resource `/payment-activations`

to recap:

* `GET /payment-requests/{codiceAvviso}
* `POST /payment-activations` -> body = `{ codice_avviso: "xyz" }`, response = `{ activation_id = "codiceContestoPagamento" }`
* `GET  /payment-activations/{activation_id}` -> body = `{ status: "PENDING" }` oppure `{ status: "COMPLETED", paymentId: "xyz" }`

Reply:
Perfect!
But I suggest to use codiceContestoPagamento instead activation_id, just to be in line with pagoPa documentation

Reply:
The first one is:
GET /payment-requests/{CF}/{codiceAvviso}

Reply:
@cloudify codiceAvviso (or better, [Numero Avviso](https://pagopa-specifichepagamenti.readthedocs.io/it/latest/_docs/Capitolo7.html?highlight=aux#il-numero-avviso-e-larchivio-dei-pagamenti-in-attesa) in PagoPA jargon, does not contain the CF of the administration, that's why is needed as an extra field in the path

Reply:
what we need here is the string provided by the QR code that contains both, I don't know if there's a name for it - if it doesn't have a name we can give one